libpgcommon/lwgeom_cache.c

```diff
@@ -236,4 +245,70 @@ GetGeomCache(FunctionCallInfo fcinfo,
 	return NULL;
 }
 
+/******************************************************************************/
+
+inline static ToastCache*
+ToastCacheGet(FunctionCallInfo fcinfo)
+{
+	const uint32_t entry_number = TOAST_CACHE_ENTRY;
+	GenericCacheCollection* generic_cache = GetGenericCacheCollection(fcinfo);
+	ToastCache* cache = (ToastCache*)(generic_cache->entry[entry_number]);
+	if (!cache)
+	{
+		cache = MemoryContextAllocZero(FIContext(fcinfo), sizeof(ToastCache));
+		cache->type = entry_number;
+		generic_cache->entry[entry_number] = (GenericCache*)cache;
+	}
+	return cache;
+}
+
+inline static uint64_t
+ToastCacheHashVarlena(const struct varlena *attr)
+{
+	struct varatt_external ve;
+	VARATT_EXTERNAL_GET_POINTER(ve, attr);
+	uint64_t a = DatumGetUInt64(hash_uint32(ve.va_valueid));
+	uint64_t b = DatumGetUInt64(hash_uint32(ve.va_toastrelid));
+	/* Merge hashes */
+	return a ^ (b + UINT64CONST(0x49a0f4dd15e5a8e3) + (a << 54) + (a >> 7));
+}
 
+GSERIALIZED*
```

Comment:
I assume this is should be a `const GSERIALIZED*` as in `thou shalt not modify this nor the geometry that you build on top of it`, right?

Reply:
Maybe so, but doing so cascade down into piles of other legacy code, so if we want to start making gserialized const over the whole code base that's a different task.